File: src/onnx9000/extensions/custom/decoding.py
"""Module providing core logic and structural definitions."""

import math
from typing import List, Callable


def greedy_search(logits: List[List[float]], eos_token_id: int = -1) -> List[int]:
    """Provides semantic functionality and verification."""
    out = []
    for step_logits in logits:
        best_idx = 0
        best_score = float("-inf")
        for j, val in enumerate(step_logits):
            if val > best_score:
                best_score = val
                best_idx = j
        out.append(best_idx)
        if best_idx == eos_token_id:
            break
    return out


def sample_top_k_top_p(
    logits: List[float], top_k: int = 0, top_p: float = 1.0, temperature: float = 1.0
) -> int:
    """Provides semantic functionality and verification."""
    import random

    if temperature != 1.0 and temperature > 0:
        logits = [l / temperature for l in logits]

    max_logit = max(logits)
    probs = [math.exp(l - max_logit) for l in logits]
    sum_probs = sum(probs)
    probs = [p / sum_probs for p in probs]

    indexed_probs = [(p, i) for i, p in enumerate(probs)]
    indexed_probs.sort(key=lambda x: x[0], reverse=True)

    filtered = indexed_probs

    if top_k > 0 and top_k < len(filtered):
        filtered = filtered[:top_k]

    if top_p < 1.0:
        cum_sum = 0.0
        cutoff_idx = len(filtered)
        for i, item in enumerate(filtered):
            cum_sum += item[0]
            if cum_sum >= top_p:
                cutoff_idx = i + 1
                break
        filtered = filtered[:cutoff_idx]

    new_sum = sum(x[0] for x in filtered)
    filtered = [(x[0] / new_sum, x[1]) for x in filtered]

    rand = random.random()
    acc = 0.0
    for p, i in filtered:
        acc += p
        if rand <= acc:
            return i

    return filtered[-1][1]


class Beam:
    """Provides semantic functionality and verification."""

    def __init__(self, seq: List[int], score: float, is_finished: bool):
        """Provides semantic functionality and verification."""
        self.seq = seq
        self.score = score
        self.is_finished = is_finished


def beam_search(
    logits_fn: Callable[[List[List[int]]], List[List[float]]],
    initial_input: List[int],
    beam_size: int,
    max_steps: int,
    eos_token_id: int = -1,
) -> List[int]:
    """Provides semantic functionality and verification."""
    beams = [Beam(list(initial_input), 0.0, False)]

    for _ in range(max_steps):
        active_beams = [b for b in beams if not b.is_finished]
        finished_beams = [b for b in beams if b.is_finished]

        if not active_beams:
            break

        seqs = [b.seq for b in active_beams]
        next_logits_batch = logits_fn(seqs)

        new_beams = []
        for b_idx, beam in enumerate(active_beams):
            logits = next_logits_batch[b_idx]

            max_l = max(logits)
            log_probs = [math.exp(l - max_l) for l in logits]
            sum_p = sum(log_probs)
            log_probs = [math.log(p / sum_p) for p in log_probs]

            indexed = [(lp, i) for i, lp in enumerate(log_probs)]
            indexed.sort(key=lambda x: x[0], reverse=True)

            for i in range(min(beam_size, len(indexed))):
                next_token = indexed[i][1]
                next_score = beam.score + indexed[i][0]
                new_beams.append(
                    Beam(
                        seq=beam.seq + [next_token],
                        score=next_score,
                        is_finished=(next_token == eos_token_id),
                    )
                )

        combined = finished_beams + new_beams
        combined.sort(key=lambda b: b.score, reverse=True)
        beams = combined[:beam_size]

    return beams[0].seq

File: src/onnx9000/extensions/custom/decoding.ts
/** Implementation details and semantic operations. */
export function greedySearch(
    logits: number[][],
    eosTokenId: number = -1
): number[] {
    const out: number[] = [];
    /** Implementation details and semantic operations. */
    for(let i = 0; i < logits.length; i++) {
        const stepLogits = logits[i];
        let bestIdx = 0;
        let bestScore = -Infinity;
        /** Implementation details and semantic operations. */
        for(let j = 0; j < stepLogits.length; j++) {
            /** Implementation details and semantic operations. */
            if(stepLogits[j] > bestScore) {
                bestScore = stepLogits[j];
                bestIdx = j;
            }
        }
        out.push(bestIdx);
        /** Implementation details and semantic operations. */
        if(bestIdx === eosTokenId) {
            break;
        }
    }
    return out;
}

/** Implementation details and semantic operations. */

export function sampleTopKTopP(
    logits: number[],
    topK: number = 0,
    topP: number = 1.0,
    temperature: number = 1.0
): number {
    /** Implementation details and semantic operations. */
    if(temperature !== 1.0 && temperature > 0) {
        /** Implementation details and semantic operations. */
        for(let i = 0; i < logits.length; i++) {
            logits[i] /= temperature;
        }
    }

    // Softmax
    let maxLogit = -Infinity;
    /** Implementation details and semantic operations. */
    for(let i = 0; i < logits.length; i++) {
        /** Implementation details and semantic operations. */
        if(logits[i] > maxLogit) maxLogit = logits[i];
    }

    let sum = 0.0;
    const probs = new Array(logits.length);
    /** Implementation details and semantic operations. */
    for(let i = 0; i < logits.length; i++) {
        probs[i] = Math.exp(logits[i] - maxLogit);
        sum += probs[i];
    }
    /** Implementation details and semantic operations. */
    for(let i = 0; i < probs.length; i++) {
        probs[i] /= sum;
    }

    const indexedProbs = probs.map((p, i) => ({ p, i }));
    indexedProbs.sort((a, b) => b.p - a.p);

    let filtered = indexedProbs;

    // Top-K
    /** Implementation details and semantic operations. */
    if(topK > 0 && topK < filtered.length) {
        filtered = filtered.slice(0, topK);
    }

    // Top-P
    /** Implementation details and semantic operations. */
    if(topP < 1.0) {
        let cumSum = 0;
        let cutoffIdx = filtered.length;
        /** Implementation details and semantic operations. */
        for(let i = 0; i < filtered.length; i++) {
            cumSum += filtered[i].p;
            /** Implementation details and semantic operations. */
            if(cumSum >= topP) {
                cutoffIdx = i + 1;
                break;
            }
        }
        filtered = filtered.slice(0, cutoffIdx);
    }

    // Renormalize
    let newSum = 0;
    /** Implementation details and semantic operations. */
    for(const item of filtered) newSum += item.p;
    /** Implementation details and semantic operations. */
    for(const item of filtered) item.p /= newSum;

    // Sample
    const rand = Math.random();
    let acc = 0;
    /** Implementation details and semantic operations. */
    for(const item of filtered) {
        acc += item.p;
        /** Implementation details and semantic operations. */
        if(rand <= acc) {
            return item.i;
        }
    }
    return filtered[filtered.length - 1].i;
}

/** Implementation details and semantic operations. */

export function beamSearch(
    logitsFn: (seqs: number[][]) => Promise<number[][]>,
    initialInput: number[],
    beamSize: number,
    maxSteps: number,
    eosTokenId: number = -1
): Promise<number[]> {
    // simplified mock of beam search algorithm
    return new Promise(async (resolve) => {
        let beams = [{ seq: [...initialInput], score: 0.0, isFinished: false }];

        /** Implementation details and semantic operations. */

        for(let step = 0; step < maxSteps; step++) {
            const newBeams: { seq: number[]; score: number; isFinished: boolean }[] = [];
            const activeBeams = beams.filter(b => !b.isFinished);
            const finishedBeams = beams.filter(b => b.isFinished);

            /** Implementation details and semantic operations. */

            if(activeBeams.length === 0) break;

            const seqs = activeBeams.map(b => b.seq);
            const nextLogitsBatch = await logitsFn(seqs);

            /** Implementation details and semantic operations. */

            for(let bIdx = 0; bIdx < activeBeams.length; bIdx++) {
                const beam = activeBeams[bIdx];
                const logits = nextLogitsBatch[bIdx];

                // Softmax log probs
                let maxLogit = -Infinity;
                /** Implementation details and semantic operations. */
                for(let i = 0; i < logits.length; i++) if (logits[i] > maxLogit) maxLogit = logits[i];

                let sum = 0;
                const logProbs = new Array(logits.length);
                /** Implementation details and semantic operations. */
                for(let i = 0; i < logits.length; i++) {
                    const prob = Math.exp(logits[i] - maxLogit);
                    sum += prob;
                    logProbs[i] = prob;
                }
                /** Implementation details and semantic operations. */
                for(let i = 0; i < logits.length; i++) {
                    logProbs[i] = Math.log(logProbs[i] / sum);
                }

                // top 2 * beamSize to give some room
                const indexed = logProbs.map((lp, i) => ({ lp, i }));
                indexed.sort((a, b) => b.lp - a.lp);

                /** Implementation details and semantic operations. */

                for(let i = 0; i < beamSize; i++) {
                    const nextToken = indexed[i].i;
                    const nextScore = beam.score + indexed[i].lp;
                    newBeams.push({
                        seq: [...beam.seq, nextToken],
                        score: nextScore,
                        isFinished: nextToken === eosTokenId
                    });
                }
            }

            const combined = [...finishedBeams, ...newBeams];
            combined.sort((a, b) => b.score - a.score);
            beams = combined.slice(0, beamSize);
        }

        /** Implementation details and semantic operations. */

        resolve(beams[0].seq);
    });
}
